data/transactions/logic/eval_test.go

```diff
@@ -3972,3 +3807,161 @@ func TestRekeyFailsOnOldVersion(t *testing.T) {
 		})
 	}
 }
+
+func obfuscate(program string) string {
+	// Put a prefix on the program that does nothing interesting,
+	// but prevents assembly from detecting type errors.  Allows
+	// evaluation testing of a program that would be rejected by
+	// assembler.
+	if strings.Contains(program, "obfuscate") {
+		return program // Already done.  Tests sometimes use at multiple levels
+	}
+	return "int 0;bnz obfuscate;obfuscate:;" + program
+}
+
+type evalTester func(pass bool, err error) bool
+
+func testEvaluation(t *testing.T, program string, introduced uint64, tester evalTester) {
+	for v := uint64(1); v <= AssemblerMaxVersion; v++ {
+		t.Run(fmt.Sprintf("v=%d", v), func(t *testing.T) {
+			if v < introduced {
+				testProg(t, obfuscate(program), v, expect{0, "...opcode was introduced..."})
+				return
+			}
+			ops := testProg(t, program, v)
+			cost, err := Check(ops.Program, defaultEvalParams(nil, nil))
+			require.NoError(t, err)
+			require.True(t, cost < 1000)
+			var txn transactions.SignedTxn
+			txn.Lsig.Logic = ops.Program
+			sb := strings.Builder{}
+			pass, err := Eval(ops.Program, defaultEvalParams(&sb, &txn))
+			ok := tester(pass, err)
+			if !ok {
+				t.Log(hex.EncodeToString(ops.Program))
+				t.Log(sb.String())
+			}
+			require.True(t, ok)
+			isNotPanic(t, err) // Never want a Go level panic.
+		})
+	}
+}
+
+func testAccepts(t *testing.T, program string, introduced uint64) {
+	testEvaluation(t, program, introduced, func(pass bool, err error) bool {
+		return pass && err == nil
+	})
+}
+func testRejects(t *testing.T, program string, introduced uint64) {
+	testEvaluation(t, program, introduced, func(pass bool, err error) bool {
+		// Returned False, but didn't panic
+		return !pass && err == nil
+	})
+}
+func testPanics(t *testing.T, program string, introduced uint64) {
+	testEvaluation(t, program, introduced, func(pass bool, err error) bool {
+		// TEAL panic! not just reject at exit
+		return !pass && err != nil
+	})
+}
+
+func TestAssert(t *testing.T) {
+	t.Parallel()
+	testAccepts(t, "int 1;assert;int 1", 3)
+	testRejects(t, "int 1;assert;int 0", 3)
+	testPanics(t, "int 0;assert;int 1", 3)
+	testPanics(t, obfuscate("assert;int 1"), 3)
+	testPanics(t, obfuscate(`byte "john";assert;int 1`), 3)
+}
+
+func TestBits(t *testing.T) {
+	t.Parallel()
+	testAccepts(t, "int 1; int 0; getbit; int 1; ==", 3)
+	testAccepts(t, "int 1; int 1; getbit; int 0; ==", 3)
```

Comment:
how about a loop over a random number with say 24 out of 64 bits set, and checking `getbit` returns a correct value for each of 64 possible positions?

(same for bytes operand)